File: src-tauri/src/tree/metadata.rs
use std::fs;
use std::time;

/// Data related to a node in a tree.
/// 
/// Although similar to [`std::fs::Metadata`](https://doc.rust-lang.org/stable/std/fs/struct.Metadata.html), 
/// this makes it easier to test.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Metadata {
    pub name: String,
    pub size: u64,
    pub permissions: fs::Permissions,
    pub modified: Option<time::SystemTime>,
    pub accessed: Option<time::SystemTime>,
    pub created: Option<time::SystemTime>
}

File: src/models/metadata.ts
interface SystemTime {
  nanos_since_epoch: number;
  secs_since_epoch: number;
}

export interface Metadata {
  name: string;
  size: number;
  modified?: SystemTime;
  accessed?: SystemTime;
  created?: SystemTime;
}

export const isMetadata = (obj: any): obj is Metadata => {
  return (
    typeof obj === "object" &&
    obj !== null &&
    "name" in obj &&
    "size" in obj &&
    typeof obj.name === "string" &&
    typeof obj.size === "number" &&
    (obj.modified === undefined || isSystemTime(obj.modified)) &&
    (obj.accessed === undefined || isSystemTime(obj.accessed)) &&
    (obj.created === undefined || isSystemTime(obj.created))
  );
};

const isSystemTime = (obj: any): obj is SystemTime => {
  return (
    typeof obj === "object" &&
    obj !== null &&
    "nanos_since_epoch" in obj &&
    "secs_since_epoch" in obj &&
    typeof obj.nanos_since_epoch === "number" &&
    typeof obj.secs_since_epoch === "number"
  );
};
